### PR TITLE
Adding 2D version of SCAF for Bijective Maps and a tutorial

### PR DESCRIPTION
**Pull request by jiangzhongshi**
Reopen of #709 
- `scaf.h/scaf.cpp` file pair implementing the bijective parameterization. Implementation is shortened from 2000~ lines to ~700
- `slim.h/slim.cpp` changes some interface of subroutines by getting rid of the `SLIMData` for reusable.
- adding an `igl::topological_fill_hole` so it can deal with surface patches with holes
- `711_SCAF/main.cpp` contain a simple demo
- Added a `camel_b.obj` mesh with boundary to showcase the parametrization.

- Add additional direct, struct-free interface for `slim` and `scaf`

### CODE REVIEW

**Comment by alecjacobson** [comment]
This is a super cool addition to libigl. One of the best papers from 2017. It will take some time to sort through the PR though, I already see a lot of potential issues. In the meantime please try to further break up the files into smaller more reusable functions and be mindful of the style guideline. I will give more specific feedback soon.

File: include/igl/surface_gradient.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#ifndef IGL_SURFACE_GRADIENT_H
+#define IGL_SURFACE_GRADIENT_H
+
+#include "igl_inline.h"
+#include <Eigen/Dense>
+#include <Eigen/Sparse>
+
+namespace igl
+{
+  IGL_INLINE void surface_gradient(const Eigen::MatrixXd &V, const Eigen::MatrixXi &F,
```
Comment (alecjacobson) [inline include/igl/surface_gradient.h]: Needs input/output documentation.

File: include/igl/surface_gradient.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+// 
+// This Source Code Form is subject to the terms of the Mozilla Public License 
+// v. 2.0. If a copy of the MPL was not distributed with this file, You can 
+// obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "surface_gradient.h"
+
+#include "grad.h"
+
+IGL_INLINE void igl::surface_gradient(const Eigen::MatrixXd &V, const Eigen::MatrixXi &F,
```
Comment (alecjacobson) [inline include/igl/surface_gradient.cpp]: break into a function that computes projection matrix and then client calls grad, build-projection, P*G

File: python/py_doc.h [hunk 3]
```diff
@@ -99,6 +99,10 @@ extern const char *__doc_igl_readTGF;
 extern const char *__doc_igl_read_triangle_mesh;
```
Comment (alecjacobson) [inline python/py_doc.h]: Please rm python related changes.

File: include/igl/harmonic.cpp [hunk 4]
```diff
@@ -70,6 +70,55 @@ IGL_INLINE bool igl::harmonic(
...
+
+  int new_vert_id = V_rows;
+  int new_face_id = real_F_num;
+
+  for (int i = 0; i < num_holes; i++, new_vert_id++)
+  {
+    int cur_bnd_size = holes[i].size();
+    int it = 0;
+    int back = holes[i].size() - 1;
+    F_filled.row(new_face_id++) << holes[i][it], holes[i][back], new_vert_id;
```
Comment (jiangzhongshi) [inline include/igl/harmonic.cpp]: Fill topological holes

File: include/igl/mapping_energy_with_jacobians.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+#ifndef IGL_MAPPING_ENERGY_WITH_JACOBIANS_H
+#define IGL_MAPPING_ENERGY_WITH_JACOBIANS_H
+
+#include "igl_inline.h"
+#include <Eigen/Dense>
+#include "MappingEnergyType.h"
+
+namespace igl
+{
+   IGL_INLINE double mapping_energy_with_jacobians(const Eigen::MatrixXd &Ji,
```
Comment (jiangzhongshi) [inline include/igl/mapping_energy_with_jacobians.h]: doc

File: include/igl/topological_hole_fill.h [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+// obtain one at http://mozilla.org/MPL/2.0/.
+#ifndef IGL_TOPOLOGICAL_HOLE_FILL_H
+#define IGL_TOPOLOGICAL_HOLE_FILL_H
+#include "igl_inline.h"
+#include <Eigen/Core>
+#include <Eigen/Sparse>
+namespace igl
+{
+    
+  // Topological fill hole on a mesh, with one additional vertex each hole
```
Comment (alecjacobson) [inline include/igl/topological_hole_fill.h]: "one additional vertex (with index `F.maxCoeff()`)"

Comment (alecjacobson) [reply include/igl/topological_hole_fill.h]: ... for each hole

File: include/igl/topological_hole_fill.h [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  //
+  template <
+  typename DerivedF,
+  typename Derivedb,
+  typename VectorIndex>
+IGL_INLINE void topological_hole_fill(
+  const Eigen::MatrixBase<DerivedF> & F,
+  const Eigen::MatrixBase<Derivedb> & b,
+  const std::vector<VectorIndex> & holes,
+  Eigen::PlainObjectBase<DerivedF> &F_filled);
```
Comment (alecjacobson) [inline include/igl/topological_hole_fill.h]: Eigen::PlainObjectBase<DerivedF> &F_filled

-->

Eigen::PlainObjectBase<DerivedF_filled> &F_filled

File: include/igl/scaf.h [hunk 8]
```diff
@@ -0,0 +1,101 @@
+// This file is part of libigl, a simple c++ geometry processing library.
+//
+// Copyright (C) 2018 Zhongshi Jiang <jiangzs@nyu.edu>
+//
+// This Source Code Form is subject to the terms of the Mozilla Public License
+// v. 2.0. If a copy of the MPL was not distributed with this file, You can
+// obtain one at http://mozilla.org/MPL/2.0/.
+#ifndef SCAF_H
```
Comment (alecjacobson) [inline include/igl/scaf.h]: IGL_SCAF_H

**Pull request merged**